include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1898,6 +1898,80 @@ SparseMatrix<number>::get_sparsity_pattern () const
 }
 
 
+namespace internal
+{
+  namespace SparseMatrix
+  {
+    template <class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const realType&val)
+    {
+      out << val;
+    }
+
+    template<class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const std::complex<realType> &val)
+    {
+      // the following format is what is required by the matrix market format.
+      out << val.real << ' ' << val.imag;
+    }
+
+    template<class StreamType, class numberType>
+    void
+    print_header(StreamType &out,const numberType & dummy)
+    {
+      out << "%%MatrixMarket matrix coordinate real general\n";
+    }
+
+    //template<typename realType>
+    template<class StreamType, class realType>
+    void
+    print_header(StreamType &out, const std::complex<realType> & dummy )
+    {
+      out << "%%MatrixMarket matrix coordinate complex general\n";
+    }  
+  } // namespace SparseMatrix
+} // namespace internal
+
+template <typename number>
+template <class StreamType>
+void SparseMatrix<number>::print_matrix_market(StreamType &out,
+					       const double threshold) const
+{
+  Assert(cols != 0, ExcNotInitialized());
+  Assert(val != 0, ExcNotInitialized());
+  Assert(threshold >= 0, ExcMessage("Negative threshold!") );  
+
+  //Print the header.
+  {
+    const number dummy = (number) 1.0; 
+    print_header(out,dummy);
```

Comment:
First of all, I am sorry this second update is still so bad.  I didn't know these changes would be visible to everyone already.

Second, please correct me if I am wrong, but without the dummy variable, it seems that nothing is begin overloaded...  I just removed the dummy variable and now the compiler complains of print_header being multiply defined.

I honestly tried to figure this out for a couple hours today. I couldn't get it to compile without completely specifying the type.  like std::complex<double> or std::complex<float> instead of std::complex<number>.  Any ideas? 

My proposal is to wait until c++11 is required then add a `static_assert(std::arithmetic<number>,"Only defined for real types");` or something similar.  This would get probably 90% of the functionality, and get rid the extra `print_value` and `print_header` functions along with the namespaces.  

Is there a tutorial that uses complex valued matrices I could test this with?

I also attempted to discern how eigen solves similar problems.  The templatize their matrices with a type but